File: 0121_Best_time_to_buy_and_sell_stock/121_best_time_to_buy_and_sell_stock.py


#time limit exceeded
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        max_profit = 0
        for i in range(len(prices)):
            for j in range(i+1, len(prices)):
                profit = prices[j] - prices[i]

                if profit > 0:
                    max_profit = max(max_profit, profit)

        return max_profit
        
        
        
time O(n) space O(1)
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        max_profit = 0
        min_price = float('inf')

        for price in prices:
            if price < min_price:
                min_price = price

            profit = price - min_price

            if profit > max_profit:
                max_profit = profit

        return max_profit

File: 0121_Best_time_to_buy_and_sell_stock/121_best_time_to_buy_and_sell_stock.c
static inline int min(int x, int y)
{
    return ((x < y) ? x : y);
}
static inline int max(int x, int y)
{
    return ((x > y) ? x : y);
}

int maxProfit(int* prices, int pricesSize) {
    int low_price = INT_MAX;
    int high_price = 0;

    for(int i=0; i<pricesSize; i++)
    {
        low_price = min(low_price, prices[i]);
        high_price = max(prices[i] - low_price, high_price);
    }

    return (high_price);
}
